### PR TITLE
feat: market data refactor to handle sorted market data IDs

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR improves both performance and heuristics after https://github.com/shapeshift/web/pull/4428 got in, sorting market data IDs (and `byId` entries) by market cap descending:

- renames `selectMarketDataSortedByMarketCap` to `selectSelectedCurrencyMarketDataSortedByMarketCap`. The name doesn't make sense anymore: the market data is now already sorted by market cap. The only thing this is really doing is massaging the market data to the selected user currency
  - simply returns the output of `selectCryptoMarketData` in case the selected currency is the default USD one
  - doesn't iterate over the `Object.entries()` of `cryptoMarketData`, but over the `marketDataAssetIds` instead, which are sorted so we can simply do a more efficient reduce over the IDs
  - switches the `.map().reduce()` call to a single `.reduce()` call
- makes `selectAssetsByMarketCap` a simplified "map over market data IDs (which are now sorted", lookup and return the `Asset` for it". While at it, removes the sorting by name, which was wrong according to the name intent. The sorting by name has been deleted, see 1dc15b1f17fe8486ed0a35e23d6d080b94ecc765 ~~This is now done in `selectAssetsByMarketCapAndName`, which itself consumes the refactored `selectAssetsByMarketCap`~~ 
- removes `getAssetMarketCap` sorting in `selectSortedAssets` and renames it to the very verbose, but very accurate `selectAssetsSortedByMarketCapFiatBalanceAndName` vernacular
- removes the unused `selectCryptoMarketDataIdsSortedByMarketCap` selector
- refactors `<TransactionHistory />`'s `useSearch` utility hook to use `selectAssets` instead of `selectAssetsByMarketCap` (we don't need any sorting there)

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Theoretically, sorting by market data could be broken all around the app - realistically, this is mostly a name change and perf. improvement, there shouldn't be any change of flow

## Testing

- Do a smoke test of assets sorting around the app
  - Asset Search in header, asset page, send/receive, trade and fiat ramps
  - DeFi section
  - Tx history (do we still need the `useSearch()` there doing any sorting at all anymore?)

### Engineering

- My renamy isn't too verbosy
- Scrutinize we are still getting the same sorted data as before

### This diff

image
image
image
image
image
image
image

### Prod

image
image
image
image
image
image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4429** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4429" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4429?utm_source=stack-comment).

File: src/state/slices/marketDataSlice/selectors.ts [hunk 1]
```diff
@@ -17,31 +17,39 @@ export const selectCryptoMarketData = (state: ReduxState) => state.marketData.cr
...
-        .map<[AssetId, MarketData]>(([assetId, assetMarketData]) => {
-          return [
-            assetId,
-            {
+        .reduce<MarketDataById<AssetId>>((acc, assetId) => {
+          const assetMarketData = cryptoMarketData[assetId]
+          // No conversion needed
+          if (selectedCurrency === 'USD') {
+            acc[assetId] = assetMarketData
+          } else {
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/state/slices/marketDataSlice/selectors.ts]: 35899e23ce512fda4162017dcda78e26bc3b34c1

File: src/state/slices/marketDataSlice/selectors.ts [hunk 2]
```diff
@@ -17,34 +17,36 @@ export const selectCryptoMarketData = (state: ReduxState) => state.marketData.cr
...
-        .reduce<MarketDataById<AssetId>>((acc, [assetId, assetMarketData]) => {
-          acc[assetId] = assetMarketData
+    // No conversion needed
+    return selectedCurrency === 'USD'
+      ? cryptoMarketData
+      : marketDataAssetIds.reduce<MarketDataById<AssetId>>((acc, assetId) => {
+          const assetMarketData = cryptoMarketData[assetId]
+          // Market data massaged to the selected currency
+          const selectedCurrencyAssetMarketData = {
+            ...assetMarketData,
```
Comment (gomesalexandre) [inline src/state/slices/marketDataSlice/selectors.ts:39]: could be mutaty for even better perf but not a big deal IMO - perhaps `Object.assign()` does make more sense however

File: src/pages/TransactionHistory/hooks/useSearch.tsx [hunk 3]
```diff
@@ -2,11 +2,11 @@ import type { Asset } from '@shapeshiftoss/asset-service'
 import debounce from 'lodash/debounce'
 import { useCallback, useState } from 'react'
 import { filterAssetsBySearchTerm } from 'components/AssetSearch/helpers/filterAssetsBySearchTerm/filterAssetsBySearchTerm'
-import { selectAssetsByMarketCap } from 'state/slices/selectors'
+import { selectAssetsByMarketCapAndName } from 'state/slices/selectors'
 import { useAppSelector } from 'state/store'
 
 export const useSearch = () => {
-  const assets = useAppSelector(selectAssetsByMarketCap)
+  const assets = useAppSelector(selectAssetsByMarketCapAndName)
```
Comment (gomesalexandre) [inline src/pages/TransactionHistory/hooks/useSearch.tsx:9]: Do we still need the `useSearch()` there doing any sorting at all here anymore? `selectAssets` without sorting seems like it is enough, given we don't really need sorting here?

Comment (gomesalexandre) [reply src/pages/TransactionHistory/hooks/useSearch.tsx]: No we don't. We don't even need assets altogether - to be tackled in a follow up. TODOs added in 1dc15b1f17fe8486ed0a35e23d6d080b94ecc765. Switched to `selectAssets` and memoized `Object.values()` of `assets` for the time being, to be updated to AssetId right on the next PR not to clutter this

File: src/state/slices/marketDataSlice/selectors.ts (hunk 2 above)
Comment (gomesalexandre) [reply src/state/slices/marketDataSlice/selectors.ts]: dcc8efbc3

File: src/pages/TransactionHistory/TransactionHistory.tsx [hunk 4]
```diff
@@ -21,6 +21,7 @@ export const TransactionHistory = () => {
   const { filters, setFilters, resetFilters } = useFilters()
   const selectorFilters = useMemo(
     () => ({
+      // TODO(gomes): This is wrong. We don't use anything Asset-y from this, only the AssetId
```
Comment (gomesalexandre) [inline src/pages/TransactionHistory/TransactionHistory.tsx:24]: revert, this is actually "fine"
```suggestion
```

File: src/pages/TransactionHistory/hooks/useSearch.tsx [hunk 5]
```diff
@@ -1,12 +1,19 @@
...
 import { useAppSelector } from 'state/store'
 
 export const useSearch = () => {
-  const assets = useAppSelector(selectAssetsByMarketCap)
+  const assetsById = useAppSelector(selectAssets)
+  // TODO(gomes): This shouldn't be a view-layer concern. To be tackled in a follow-up PR not to tackle the diff of https://github.com/shapeshift/web/pull/4429
+  // - selectAssets will be LSP renamed to selectAssetsById
+  // - selectAssets will now be a selector that returns Asset[]
+  // We could use selectAssetsByMarketCap but this is wrong. We don't care about the market cap here.
+  // While at it, this shouldn't be selecting assets altogether, but AssetIds, see consumption in <TransactionHistory />
```
Comment (gomesalexandre) [inline src/pages/TransactionHistory/hooks/useSearch.tsx:15]: Remove TODO. Not worth the rename hassle for a one-off use-case, we really don't need unsorted `Asset[]` except here
```suggestion
```

**Review by 0xApotheosis — APPROVED** [review]
Tidy refactor, I'm all for the explicit naming 👌 

Regression tested sorting against prod and found no delta ✅

**Pull request merged**